collection_mutation.hh

```diff
@@ -0,0 +1,121 @@
+    utils::chunked_vector<std::pair<bytes_view, atomic_cell_view>> cells;
+
+    // Copies the observed data, storing it in a collection_mutation_description.
+    collection_mutation_description materialize(const abstract_type&) const;
+
+    // Packs the data to a serialized blob.
+    collection_mutation serialize(const abstract_type&) const;
+};
+
+// Given a linearized collection_mutation_view, returns an auxiliary struct allowing the inspection of each cell.
+// The struct is an observer of the data given by the collection_mutation_view and doesn't extend its lifetime.
+// The function needs to be given the type of stored data to reconstruct the structural information.
+collection_mutation_view_description deserialize_collection_mutation(const abstract_type&, bytes_view);
+
+class collection_mutation_view {
+public:
+    atomic_cell_value_view data;
+
+    // Is this a noop mutation?
+    bool is_empty() const;
+
+    // Is any of the stored cells live (not deleted nor expired) at the time point `tp`,
+    // given the later of the tombstones `t` and the one stored in the mutation (if any)?
+    // Requires a type to reconstruct the structural information.
+    bool is_any_live(const abstract_type&, tombstone t = tombstone(), gc_clock::time_point tp = gc_clock::time_point::min()) const;
+
+    // The maximum of timestamps of the mutation's cells and tombstone.
+    api::timestamp_type last_update(const abstract_type&) const;
+
+    // Given a function that operates on a collection_mutation_view_description,
+    // calls it on the corresponding description of `this`.
+    template <typename F>
+    inline decltype(auto) with_deserialized(const abstract_type& type, F f) const {
+        return data.with_linearized([&] (bytes_view bv) {
+            return f(deserialize_collection_mutation(type, std::move(bv)));
+        });
+    }
+};
+
+// A serialized mutation of a collection of cells.
+// Used to represent mutations of collections (lists, maps, sets) or non-frozen user defined types.
+// It contains a sequence of cells, each representing a mutation of a single entry (element or field) of the collection.
+// Each cell has an associated 'key' (or 'path'). The meaning of each (key, cell) pair is:
+//  for sets: the key is the serialized set element, the cell contains no data (except liveness information),
+//  for maps: the key is the serialized map element's key, the cell contains the serialized map element's value,
+//  for lists: the key is a timeuuid identifying the list entry, the cell contains the serialized value,
+//  for user types: the key is an index identifying the field, the cell contains the value of the field.
+//  The mutation may also contain a collection-wide tombstone.
+class collection_mutation {
+public:
+    using imr_object_type =  imr::utils::object<data::cell::structure>;
+    imr_object_type _data;
+
+    collection_mutation() {}
+    collection_mutation(const abstract_type&, collection_mutation_view);
```

Comment:
We discussed this with Kamil quite a lot during the development of this patch. If I remember correctly, initially, Kamil has changed the name but then it turned out that this object is a mutation of collection of cells. Both UDTs and tuples if are not frozen are a collection of cells and fits this as well. In the end, we were not able to come up with a better name to describe a mutation of collection of cells. Would you have any suggestions @espindola?

Reply:
Actually I'd be glad to give it a better name.
As @haaawk said it is a "mutation of a collection of cells", but it's still easy to get confused by someone who's reading some code involving collection_mutations for the first time and assume that they only deal with collection types (fortunately, the doc comment explains that it involves UDTs too).

We already have names for the set of non-frozen collections and non-frozen UDTs: "non-atomic" or "multi-cell". So I thought about renaming collection_mutation to multi_cell_mutation. But then I didn't know what to do with "atomic_cell_or_collection", which is a variant<atomic_cell, collection_mutation> :D 
(actually, atomic_cell_or_collection could probably be called column_mutation, because it represents a mutation of a single column entry in a single row, if I understand correctly).

Reply:
Why call it a mutation in the first place? A mutation is Scylla's word for a database row - in the sense that it is a LSM representation of a database row. This class applies to one cell, not an entire row.

Reply:
@kostja I don't think mutation means database row in Scylla. It very often means the whole partition. In fact mutation class defined in mutation.hh represents the mutation of the whole partition.

Reply:
You can see from the fact that we have this discussion the term is abused. Let's have partition_mutation, row_mutation and cell_mutation as base terms. Then we can have collection_cell_mutation as a class name. But that would require consistent application in other domains in the source code.

Reply:
> collection_cell_mutation
more like collection_of_cells_mutation, or just cell_collection_mutation.

Here's how I understand all this stuff:

a cell mutation, or simply a cell, is the smallest (atomic) possible change (modification, or mutation) you can make to the database. It contains liveness information (TTL, timestamp, tombstone, etc.), and most of the time, additional data.

Mutations can be composed to obtain "bigger" mutations. The composition operation should be associative (`(a . b) . c == a . (b . c)`). Mutations should also be idempotent (`a . a = a`). Simple algebraic stuff.

Confusion arises when we try to mix the algebraic mutation stuff with the DB terminology (columns, rows, partitions, tables). A database is just a (possibly very big) mutation. Mutations can be applied to a different database (i.e., composed with) to obtain a modified (mutated) database.

In our code, the word "cell" is overloaded. One time it means the smallest possible mutation (see the `atomic_cell.hh` header), in other place it means `atomic_cell_or_collection`, or in other words, a mutation of a single column in a single row:
`void set_cell(const clustering_key_prefix& prefix, const column_definition& def, atomic_cell_or_collection&& value);` (from `mutation.hh`).
So a mutation of a column is either an `atomic_cell` or a collection of atomic cells, i.e. `collection_mutation`. This variant is represented as `atomic_cell_or_collection`. The first case is used for "atomic types" (int, text, frozen map, frozen UDT, etc.), and the other for "non-atomic" types (map, set, list, non-frozen UDT).

When `atomic_cell_or_collection`s are gathered to form a row, they are called `row` (see mutation_partition.hh). It could be called row_mutation. Then `row` is wrapped into a `deletable_row`, which is then wrapped into a `rows_entry`, and a bunch of such entries is called `mutation_partition` (should be `partition_mutation` IMO, because it's really a "mutation of a partition").

The codebase unfortunately often mixes interface with implementation: `atomic_cell_or_collection` is calling the class based on its implementation. The interfacey way to call it would probably be `column_mutation`, or just `column`.
On the other hand, `row` is calling the class based on interface. The implementation way would be calling it `cell_set` or `cell_sequence` or whatever.

Reply:
I am still fairly new, so I don't think I can suggest a good name, other than something super generic like change_list.
I am OK with any name really, sorry for the bikeshedding.

Reply:
Let's leave any further name changes, if any, for the future.

Reply:
It's somewhat problematic in that the usual collections are homogeneous while tuple/UDT are heterogeneous. Collections have unlimited size, while tuple/UDT have fixed size (that can be altered, but for a particular schema, it is fixed).

Collections are ordered by their values, tuples are ordered by position.

Right now it works because you view a tuple as a map<index, blob> but as soon as we want to make use of individual types, it will break apart.

I have a feeling this merging of tuples and collections is accidental and we'll need to split it in the future into two classes.

Reply:
In that case it should be an easy refactor (much easier than what I did here).